### PR TITLE
Adding SPARCNet and ContraWR

### PR DESCRIPTION
**Pull request by bruAristimunha**
In Biot's article, in addition to the BIOT model, the author centralizes codes from other models.

The list is:

1. SPaRCNet,
2. ContraWR,
3. CNN-Transformer,
4. FFCL and
5. ST-Transformer (predecessor of EEG Conformer, which we already have implemented here).

In this PR, I am proposing to implement 1) SPaRCNet and 2) ContraWR. 

3) CNN-Transformer seemed very similar to ContraWR, I don't think it's worth implementing.
4) The FFCL does not appear to be correctly implemented.

### CODE REVIEW

**Comment by bruAristimunha** [comment]
fyi @mccorsi

File: braindecode/models/sparcnet.py [hunk 1]
```diff
@@ -0,0 +1,353 @@
...
+    --------
+    >>> x = torch.randn(128, 5, 1000)
+    >>> model = TransitionLayer(5, 18)
+    >>> y = model(x)
+    >>> y.shape
+    torch.Size([128, 18, 500])
+    """
+
+    def __init__(
+        self, input_channels, output_channels, conv_bias=True, batch_norm=True
```
Comment (agramfort) [inline braindecode/models/sparcnet.py:171]: you should aim for consistency with pytorch naming conventions as much as possible so there use in_channels, out_channels

File: braindecode/models/sparcnet.py [hunk 2]
```diff
@@ -0,0 +1,353 @@
...
+    This implementation is not guaranteed to be correct, has not been checked
+    by original authors.
+
+    Parameters
+    ----------
+    block_layers : int, optional
+        Number of layers per dense block. Default is 4.
+    growth_rate : int, optional
+        Growth rate of the DenseNet. Default is 16.
+    bn_size : int, optional
```
Comment (agramfort) [inline braindecode/models/sparcnet.py:211]: bn sounds like batch_norm on first read. maybe use bottleneck_size to be explicit

File: braindecode/models/sparcnet.py [hunk 3]
```diff
@@ -0,0 +1,353 @@
...
+            n_times=n_times,
+            input_window_seconds=input_window_seconds,
+            sfreq=sfreq,
+        )
+        del n_outputs, n_chans, chs_info, n_times, sfreq, input_window_seconds
+
+        # add initial convolutional layer
+        # the number of output channels is the smallest power of 2
+        # that is greater than the number of input channels
+        out_channels = 2 ** (math.floor(np.log2(self.n_chans)) + 1)
```
Comment (agramfort) [inline braindecode/models/sparcnet.py:263]: why not use math.log2 here as you use math.floor

**Review by agramfort — COMMENTED** [review]
at some point it would be good to do a full evaluation of all these architectures on a some standard datasets

**Comment by bruAristimunha** [comment]
The green button is yours @agramfort

**Pull request merged**